specs/PushNotifications/PushNotifications-spec.md

```diff
@@ -0,0 +1,378 @@
+
+        auto payload = pushArgs.Payload();
+
+        // Do stuff to process the raw payload
+
+        // Call CompleteDeferral as good practise: Needed mainly for low power usage
+        pushArgs.CompleteDeferral();
+    }
+    else if (kind == ReunionActivationKind::Launch) // This indicates that the app is launching in the foreground
+    {
+        // Register an event to Intercept Push payloads
+        auto eventToken = details.PushActivated([](const auto&, PushActivatedEventArgs args)
+        {
+                // Call TakeDeferral to ensure that code runs in low power
+                args.GetDeferral();
+
+                auto payload = args.Payload();
+
+                // Do stuff to process the raw payload
+
+                // Call CompleteDeferral as good practise: Needed mainly for low power usage
+                args.CompleteDeferral();
+        });
+
+        // {F80E541E-3606-48FB-935E-118A3C5F41F4}
+        winrt::guid remoteId{ 0xf80e541e, 0x3606, 0x48fb, {0x93, 0x5e, 0x11, 0x8a, 0x3c, 0x5f, 0x41, 0xf4} };
+        auto channelOperation = PushManager::CreateChannelAsync(remoteId);
+
+        // Setup the inprogress event handler
+        channelOperation.Progress(
+            [](
+                IAsyncOperationWithProgress<ChannelResult, ChannelResult> const& /* sender */,
+                ChannelResult const& args)
+            {
+                if (args.Status() == ChannelStatus::InProgress)
+                {
+                    // This is basically a noop since it isn't really an error state
+                    printf("The first channel request is still in progress! \n");
+                }
+                else if (args.Status() == ChannelStatus::InProgressRetry)
+                {
+                    LOG_HR_MSG(args.ExtendedError(), "The channel request is in back-off retry mode because of a retryable error! Expect delays in acquiring it.");
+                }
+            });
+
+        // Setup the completed event handler
+        channelOperation.Completed(
+            [](
+                IAsyncOperationWithProgress<ChannelResult, ChannelResult> const& sender,
+                AsyncStatus const /* asyncStatus */)
+            {
+                auto result = sender.GetResults();
+                if (result.Status() == ChannelStatus::CompletedSuccess)
+                {
+                    auto channelUri = result.Channel().Uri();
+                    auto channelExpiry = result.Channel().ExpirationTime();
+
+                    // Persist the channelUri and Expiry in the App Service
+                }
+                else if (result.Status() == ChannelStatus::CompletedFailure)
+                {
+                    LOG_HR_MSG(result.ExtendedError(), "We hit a critical non-retryable error with channel request!");
+                }
+            });
+
+        // Draw window and other foreground UI stuff here
+
+        // Unsubscribe the foreground event
+        details.PushActivated(eventToken);
```

Comment:
How does this unsubscribe? Maybe the comment (or the API call) is incorrect? Something seems wrong here ;)

Reply:
@andrewleader Subscription returns an eventToken in c++/winrt. See

auto eventToken = details.PushActivated(eventhandler).

The corresponding equivalent to unregister the eventtoken is to pass the token back to PushActivated for unsubscribe operation:
auto eventToken = details.PushActivated


The generated function calls look like this:

struct InProcActivatorDetails : InProcActivatorDetailsT<InProcActivatorDetails>
    {
        ********
        winrt::event_token PushActivated(Windows::Foundation::EventHandler<Microsoft::ProjectReunion::PushActivatedEventArgs> const& handler);
        void PushActivated(winrt::event_token const& token) noexcept;
*******
    };

Reply:
Ah, I guess that's just how C++/WinRT works with event handlers, funky! Oh well, seems good then!